### PR TITLE
Add basic http-get and authenticate request handlers

### PR DESCRIPTION
**Pull request by Jharaxus**
- Adds architecture for handling http get requests by the server.
- Adds basic handling of authenticate requests from a client

### CODE REVIEW

**Comment by Jharaxus** [comment]
Reopened after adding latest specification changes.

File: be2-scala/src/main/scala/ch/epfl/pop/authentication/Authenticate.scala [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+      _ <- verifyResponseType(params.response_type)
+      _ <- verifyScope(params.scope)
+      _ <- verifyRedirectUri(params.redirect_uri)
+      result <- verifyResponseMode(params.response_mode)
+    } yield result
+  }
+
+  private def verifyResponseType(response_type: String): VerificationState = {
+    val expectedResponseType = "id_token token"
+    if (response_type == expectedResponseType) Right(())
```
Comment (K1li4nL) [inline be2-scala/src/main/scala/ch/epfl/pop/authentication/Authenticate.scala:58]: nitpick alert: I think a line return would make it more readable. 
```suggestion
    if (response_type == expectedResponseType) 
       Right(())
```

File: be2-scala/src/main/scala/ch/epfl/pop/authentication/Authenticate.scala [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    else
+      Left("unsupported_response_type" -> s"expected \"$expectedResponseType\" but received \"$response_type\"")
+  }
+
+  private def verifyScope(scope: String): VerificationState = {
+    val expectedScope = "openid"
+    val scopes = scope.split(" ")
+    if (scopes.contains(expectedScope)) Right(())
```
Comment (K1li4nL) [inline be2-scala/src/main/scala/ch/epfl/pop/authentication/Authenticate.scala:66]: nitpick alert: I think a line return would make it more readable. 
```suggestion
    if (scopes.contains(expectedScope))
       Right(())
```

File: be2-scala/src/main/scala/ch/epfl/pop/authentication/Authenticate.scala [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    else
+      Left("invalid_scope" -> s"expected scope to contain \"$expectedScope\" but received \"$scope\"")
+  }
+
+  private def verifyRedirectUri(uri: String): VerificationState = {
+    // Checks that the uri is a http or https format uri
+    // see https://stackoverflow.com/questions/3809401/what-is-a-good-regular-expression-to-match-a-url for regex source
+    val httpRegex =
+      "https?:\\/\\/(www\\.)?[-a-zA-Z0-9@:%._\\+~#=]{1,256}\\.[a-zA-Z0-9()]{1,6}\\b([-a-zA-Z0-9()@:%_\\+.~#?&//=]*)"
+    if (uri.matches(httpRegex)) Right(())
```
Comment (K1li4nL) [inline be2-scala/src/main/scala/ch/epfl/pop/authentication/Authenticate.scala:76]: nitpick alert: I think a line return would make it more readable. 
```suggestion
    if (uri.matches(httpRegex))
       Right(())
```

File: be2-scala/src/main/scala/ch/epfl/pop/authentication/Authenticate.scala [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    else
+      Left("invalid_request" -> s"expected http or https url format for redirect uri")
+  }
+
+  private def verifyResponseMode(mode: Option[String]): VerificationState = {
+    val supportedModes = List("query", "fragment")
+    if (mode.isEmpty || supportedModes.contains(mode.get)) Right(())
```
Comment (K1li4nL) [inline be2-scala/src/main/scala/ch/epfl/pop/authentication/Authenticate.scala:83]: nitpick alert: I think a line return would make it more readable. 
```suggestion
    if (mode.isEmpty || supportedModes.contains(mode.get)) 
      Right(())
```

File: be2-scala/src/main/scala/ch/epfl/pop/authentication/Authenticate.scala [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        "response_type",
+        "client_id",
+        "redirect_uri",
+        "scope",
+        "state".optional,
+        "response_mode".optional,
+        "login_hint",
+        "nonce"
+      ) {
+        (response_type, client_id, redirect_uri, scope, state, response_mode, login_hint, nonce) =>
```
Comment (K1li4nL) [inline be2-scala/src/main/scala/ch/epfl/pop/authentication/Authenticate.scala:28]: I think `client_id`,  `state`, `login_hint` and the `nonce` are left unchecked, wouldn't be a good idea to do so (if possible) or is that something bound to be done later maybe ?

File: be2-scala/src/main/scala/ch/epfl/pop/authentication/Authenticate.scala
Comment (K1li4nL) [inline be2-scala/src/main/scala/ch/epfl/pop/authentication/Authenticate.scala:1]: I  think this file deserve some more documentation / comments in general

File: be2-scala/src/test/scala/ch/epfl/pop/authentication/AuthenticateSuite.scala [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+import org.scalatest.funsuite.AnyFunSuite
+import org.scalatest.matchers.should.Matchers
+
+class AuthenticateSuite extends AnyFunSuite with Matchers with ScalatestRouteTest {
+
+  private val routeName = "authenticate/"
+
+  private val responseType = "id_token token"
+  private val clientID = "abc"
+  private val redirectUri = "https://wikipedia.org"
```
Comment (K1li4nL) [inline be2-scala/src/test/scala/ch/epfl/pop/authentication/AuthenticateSuite.scala:16]: nitpick: I think the domain `example.com` exists for these use case, though I don't think Wikipedia would come back at us...

File: be2-scala/src/test/scala/ch/epfl/pop/authentication/AuthenticateSuite.scala [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+      val subParams = mandatoryParams.filter(_._1 != name)
+      val request = buildRequest(subParams)
+
+      request ~> route ~> check {
+        status shouldBe Found
+        getAttributeValue(response, "error") shouldBe Some("invalid_request")
+        getAttributeValue(response, "error_description") shouldBe Some(s"Missing parameters: [$name]")
+      }
+    }
+  }
```
Comment (K1li4nL) [inline be2-scala/src/test/scala/ch/epfl/pop/authentication/AuthenticateSuite.scala:140]: A lot seems to be going there, I am not sure to follow, some comments would help ;)

**Review by K1li4nL — CHANGES_REQUESTED** [review]
Overall the code is concise and good looking, well done ! 
Beside my nitpicks, I think some documentation would be nice ;)

File: be2-scala/src/main/scala/ch/epfl/pop/authentication/Authenticate.scala (hunk 5 above)
Comment (Jharaxus) [reply be2-scala/src/main/scala/ch/epfl/pop/authentication/Authenticate.scala:28]: It is actually intended as they are left for the client (github for example using popcha for authentication) to use: client_id, state (optional) and nonce are entirely client-defined --> there are no ways to check them, neither any properties that we would like to enforce (nonce has a client-side property only). State actually should be a valid lao_id, but this doesn't need to be checked here as the user's pop app will do it to provide better flexibility, and we wouldn't actually have any means to check that now without heavily breaking the current encapsulation anyways (but again it is intended not to make that check here).

File: be2-scala/src/main/scala/ch/epfl/pop/authentication/Authenticate.scala
Comment (Jharaxus) [reply be2-scala/src/main/scala/ch/epfl/pop/authentication/Authenticate.scala:1]: Done! I commented all the public scope methods/classes and refactored a bit some parts to make them clear enough without comments.

File: be2-scala/src/test/scala/ch/epfl/pop/authentication/AuthenticateSuite.scala (hunk 6 above)
Comment (Jharaxus) [reply be2-scala/src/test/scala/ch/epfl/pop/authentication/AuthenticateSuite.scala:16]: Wow I didn't know that domain actually existed for real 😳

File: be2-scala/src/test/scala/ch/epfl/pop/authentication/AuthenticateSuite.scala (hunk 7 above)
Comment (Jharaxus) [reply be2-scala/src/test/scala/ch/epfl/pop/authentication/AuthenticateSuite.scala]: I simplified the code a bit and changed some variables names to make it clearer

**Review by K1li4nL — APPROVED** [review]
LGTM !

File: be2-scala/src/main/scala/ch/epfl/pop/authentication/Authenticate.scala [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+      result <- verifyResponseMode(params.response_mode)
+    } yield result
+  }
+
+  private def verifyResponseType(response_type: String): VerificationState = {
+    val expectedResponseType = "id_token token"
+    if (response_type == expectedResponseType)
+      Right(())
+    else
+      Left("unsupported_response_type" -> s"expected \"$expectedResponseType\" but received \"$response_type\"")
```
Comment (Ajkunas) [inline be2-scala/src/main/scala/ch/epfl/pop/authentication/Authenticate.scala:70]: It is a general comment: I suggest that you declare the type of errors as constant values. 
You can then reuse them in your test suites.

**Review by Ajkunas — CHANGES_REQUESTED** [review]
Very clean and concise code :) We easily understand what happens in the file. Still do not hesitate to add some comments here and there. 
Otherwise, nothing really important to add/change. I left a suggestion and I am open to discuss it :)

File: be2-scala/src/test/scala/ch/epfl/pop/authentication/AuthenticateSuite.scala [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+      clientID,
+      redirectUri,
+      scope,
+      "",
+      "",
+      loginHint,
+      nonce
+    )
+
+    println(request)
```
[the remark this answers is not in the document]
Comment (Jharaxus) [reply be2-scala/src/test/scala/ch/epfl/pop/authentication/AuthenticateSuite.scala:74]: Indeed, I will remove it !

File: be2-scala/src/main/scala/ch/epfl/pop/authentication/Authenticate.scala (hunk 8 above)
Comment (Jharaxus) [reply be2-scala/src/main/scala/ch/epfl/pop/authentication/Authenticate.scala]: That makes sense! I put them directly in the Authenticate object as they are not enough general-purpose to justify a separate file for them (and there are only 3).

**Review by Ajkunas — APPROVED** [review]
LGTM

**Pull request merged**